### ISSUE TITLE
Potential incomplete search in cuckoohash_map::slot_search?

### ISSUE THREAD

**Issue by yonch** [efficient/libcuckoo#90]
tl;dr: is slot_search terminating too early, wasting work filling its queue?

slot_search is meant to perform a BFS search of a certain depth (MAX_BFS_PATH_LEN) from two initial buckets, to find an empty slot. The search proceeds by enqueuing into a queue called 'q', and performing a while loop dequeuing from the queue. The size of the queue is limited to b_queue::MAX_CUCKOO_COUNT elements.

The loop currently terminates on two conditions:
1. q.empty(). Reasonable, as it would happen when the function exhausts the search depth
2. q.full(). This is why I'm opening this issue. Doing this check means the function did all the work to get results into the queue, and then stops before looking at all of these.

The for loop iterating over slots already checks q.full(), so it seems the q.full() check in the while loop can be removed. This would search to the full specified depth, while still using the queue to limit the search space -- was this the original intention?

**Comment by manugoyal** [comment]
Yes this is a great point. I think it's generally unnecessary to limit the search by `MAX_CUCKOO_COUNT` anyways, and just allocate enough space to fit a maximal `slot_search` in the queue. e0eeb904a469816b6a6c43d01ff03313ab42500a implements this, and a few other adjustments to the `slot_search` function.

Any sort of change like this would be expected to decrease performance slightly (for longer `slot_search` runtimes), but with the benefit of increased table load factors. And this is indeed what we see with this commit. IMO the small insert performance drop is justified by the more significant load-factor increase. Thanks for bringing this up!
